### ISSUE TITLE
Sporadic(?) EphemerisDao Test Error

### ISSUE THREAD

**Issue by swalker2m** [gemini-hlsw/gem#241]
```
[info] - EphemerisDao should selectAll *** FAILED ***
[info]   BatchUpdateException was thrown during property evaluation.
[info]     Message: Batch entry 0 
[info]             INSERT INTO ephemeris (key_type,
[info]                                    key,
[info]                                    site,
[info]                                    timestamp,
[info]                                    ra,
[info]                                    dec,
[info]                                    ra_str,
[info]                                    dec_str,
[info]                                    delta_ra,
[info]                                    delta_dec)
[info]                  VALUES ('Comet', 'jxd6mdgafn', 'GS', '1301-02-29 22:13:15.805159+00 BC'::timestamp with time zone, 14454775807, 0, '04:00:54.775807', '+00:00:00.000000', '-4.923000', '-8.245000')
[info]            was aborted: ERROR: date/time field value out of range: "1301-02-29 22:13:15.805159+00 BC"  Call getNextException to see other errors in the batch.
[info]     Occurred when passed generated values (
[info]       arg0 = KS(AsteroidNew(ocyh5sijty),GN),
[info]       arg1 = Ephemeris(Map(+70322-12-23T17:25:28.794547Z -> EphemerisCoordinates(Coordinates(00:00:00.000000 +31:59:05.224193),Offset(P(Angle(DMS(359:59:59.999000), 359.9999997222°)),Q(Angle(DMS(00:00:10.000000), 0.0027777778°)))), +97332-06-14T13:27:23.987179Z -> EphemerisCoordinates(Coordinates(00:00:00.000000 +00:00:00.000000),Offset(P(Angle(DMS(00:00:10.000000), 0.0027777778°)),Q(Angle(DMS(00:00:04.505000), 0.0012513889°)))), +136893-01-02T10:43:12.441564Z -> EphemerisCoordinates(Coordinates(00:00:00.000000 +31:59:05.224193),Offset(P(Angle(DMS(00:00:07.408000), 0.0020577778°)),Q(Angle(DMS(359:59:50.000000), 359.9972222222°)))), +143619-06-02T04:52:12.739696Z -> EphemerisCoordinates(Coordinates(00:00:00.000000 +00:00:00.000000),Offset(P(Angle(DMS(359:59:50.987000), 359.9974963889°)),Q(Angle(DMS(359:59:52.603000), 359.9979452778°)))), +153873-10-22T13:24:38.865391Z -> EphemerisCoordinates(Coordinates(04:00:54.775807 +00:00:00.000000),Offset(P(Angle(DMS(00:00:10.000000), 0.0027777778°)),Q(Angle(DMS(359:59:50.000000), 359.9972222222°)))), +174549-10-20T07:58:36.077247Z -> EphemerisCoordinates(Coordinates(00:00:00.000000 +31:59:05.224193),Offset(P(Angle(DMS(00:00:01.230000), 0.0003416667°)),Q(Angle(DMS(359:59:59.999000), 359.9999997222°)))), +196973-10-30T06:11:42.373376Z -> EphemerisCoordinates(Coordinates(00:00:00.000000 -31:59:05.224192),Offset(P(Angle(DMS(359:59:54.546000), 359.9984850000°)),Q(Angle(DMS(359:59:54.933000), 359.9985925000°)))), +233371-04-25T09:13:24.891094Z -> EphemerisCoordinates(Coordinates(00:00:00.000000 +00:00:00.000000),Offset(P(Angle(DMS(00:00:01.095000), 0.0003041667°)),Q(Angle(DMS(00:00:07.614000), 0.0021150000°)))), +245373-01-14T21:46:42.108485Z -> EphemerisCoordinates(Coordinates(04:00:54.775807 +31:59:05.224193),Offset(P(Angle(DMS(00:00:00.000000), 0.0000000000°)),Q(Angle(DMS(00:00:00.000000), 0.0000000000°)))))),
[info]       arg2 = Map(KS(Comet(jxd6mdgafn),GS) -> Ephemeris(Map(-1300-02-17T22:13:15.805159Z -> EphemerisCoordinates(Coordinates(04:00:54.775807 +00:00:00.000000),Offset(P(Angle(DMS(359:59:55.077000), 359.9986325000°)),Q(Angle(DMS(359:59:51.755000), 359.9977097222°)))), +16330-05-25T21:31:19.567588Z -> EphemerisCoordinates(Coordinates(04:00:54.775807 +00:00:00.000000),Offset(P(Angle(DMS(359:59:50.000000), 359.9972222222°)),Q(Angle(DMS(00:00:00.000000), 0.0000000000°)))), +90542-06-02T06:19:51.039586Z -> EphemerisCoordinates(Coordinates(00:00:00.000000 +31:59:05.224193),Offset(P(Angle(DMS(00:00:10.000000), 0.0027777778°)),Q(Angle(DMS(359:59:50.000000), 359.9972222222°)))), +140462-05-24T04:14:03.170839Z -> EphemerisCoordinates(Coordinates(19:59:05.224192 -31:59:05.224192),Offset(P(Angle(DMS(00:00:08.487000), 0.0023575000°)),Q(Angle(DMS(359:59:53.695000), 359.9982486111°))))))) // 6 shrinks
[info]     )

```

**Comment by tpolecat** [comment]
Curious. I wonder if we're parsing Horizons dates wrong.

**Comment by tpolecat** [comment]
Oh we must be generating ridiculous dates in the tests.

**Comment by swalker2m** [comment]
Yes, but I thought `Timestamp` was going to solve that problem so I must have overlooked something.

**Comment by swalker2m** [comment]
> 1301-02-29

The date is in range for Postgres, but I assume (proleptic) 1301 BC wasn't a leap year.

**Comment by tpolecat** [comment]
Haha I didn't notice that. You should file a bug report with Horizons.

**Comment by swalker2m** [comment]
Somewhere there is a problem with converting `Timestamp` into a value to be inserted into the database.  The arbitrary itself seems fine:

```scala
  implicit val arbTimestamp: Arbitrary[Timestamp] =
    Arbitrary {
      for {
        m <- Gen.choose(0L, Duration.between(Timestamp.Min.toInstant, Timestamp.Max.toInstant).toMillis)
        u <- Gen.choose(0, 999L)
      } yield Timestamp.Min.plusMillis(m).flatMap(_.plusMicros(u)).getOrElse(Timestamp.Min)
    }
```

It just picks an amount of time between 0 and the total time between the min and max values and adds that to the minimum value.  So this should produce a valid `Timestamp` and yet we try to write a date on February 29 in a non-leap year (1301 BC):

```sql
gem=# select '1301-02-29 22:13:15.805159+00 BC'::timestamp with time zone;
ERROR:  date/time field value out of range: "1301-02-29 22:13:15.805159+00 BC"
LINE 1: select '1301-02-29 22:13:15.805159+00 BC'::timestamp with ti...
```

1305 BC was a leap year (assuming you extend the Gregorian calendar back as though it were in effect at the time) so this works:

```sql
gem=# select '1305-02-29 22:13:15.805159+00 BC'::timestamp with time zone;
              timestamptz               
----------------------------------------
 1305-02-29 17:30:29.805159-04:42:46 BC
(1 row)
```

Although the timezone adjustment of `-04:42:46` seems weird.

**Comment by swalker2m** [comment]
Actually, I just noticed that date we generated seems to be: `-1300-02-17T22:13:15.805159Z` which is 1301-02-17 BC, a perfectly cromulent date.

```
arg2 = Map(KS(Comet(jxd6mdgafn),GS) -> Ephemeris(Map(-1300-02-17T22:13:15.805159Z ...
```

Poking around a bit more:

```
scala> Instant.parse("-1300-02-17T22:13:15.805159Z")
res4: java.time.Instant = -1300-02-17T22:13:15.805159Z

scala> java.sql.Timestamp.from(res4)
res5: java.sql.Timestamp = 1301-02-29 18:13:15.805159
```

BOOM!

**Comment by swalker2m** [comment]
Closed via #246.
